pkg/uefi/nvram.go

```diff
@@ -0,0 +1,458 @@
+				return fmt.Errorf("extended header size (0x%X) is too small for hash", extendedHeaderSize)
+			}
+			v.Hash = v.buf[hashstart : hashstart+sha256.Size]
+		}
+		knownExtDataFormat = true
+	}
+	v.UnknownExtendedHeaderFormat = !knownExtDataFormat
+	return nil
+}
+
+func (v *NVar) parseDataOnly(s *NVarStore) bool {
+	if v.Header.Attributes&NVarEntryDataOnly == 0 {
+		return false
+	}
+	// Search previously added entries for a link to this variable
+	var link *NVar
+	for _, l := range s.Entries {
+		if l.IsValid() && l.NextOffset == v.Offset {
+			link = l
+			break
+		}
+	}
+	if link != nil {
+		v.GUID = link.GUID
+		v.Name = link.Name
+		if v.NextOffset == 0 {
+			v.Type = DataNVarEntry
+		}
+	} else {
+		v.Name = "Invalid link"
+		v.Type = InvalidLinkNVarEntry
+	}
+
+	return true
+}
+
+func (v *NVar) parseGUID(s *NVarStore) error {
+	r := bytes.NewReader(v.buf[v.DataOffset:])
+	if v.Header.Attributes&NVarEntryGUID != 0 {
+		// GUID in variable
+		if err := binary.Read(r, binary.LittleEndian, &v.GUID); err != nil {
+			return err
+		}
+		v.DataOffset += int64(binary.Size(v.GUID))
+	} else {
+		// GUID index in store
+		var guidIndex uint8
+		if err := binary.Read(r, binary.LittleEndian, &guidIndex); err != nil {
+			return err
+		}
+		v.GUIDIndex = &guidIndex
+		v.GUID = s.getGUIDFromStore(guidIndex)
+		v.DataOffset += int64(binary.Size(guidIndex))
+	}
+	return nil
+}
+
+func (v *NVar) parseName() error {
+	if v.Header.Attributes&NVarEntryASCIIName != 0 {
+		// Name is stored as ASCII string of CHAR8s
+		namebuf := v.buf[v.DataOffset:]
+		end := bytes.IndexByte(namebuf, 0)
+		if end == -1 {
+			return io.EOF
+		}
+		v.Name = string(namebuf[:end])
+		v.DataOffset += int64(end) + 1
+	} else {
+		// Name is stored as UCS2 string of CHAR16s
+		namebuf := v.buf[v.DataOffset:]
+		end := bytes.Index(namebuf, []byte{0, 0})
+		if end == -1 {
+			return io.EOF
+		}
+		v.Name = unicode.UCS2ToUTF8(namebuf[:end])
+		v.DataOffset += int64(end) + 2
+	}
+	return nil
+}
+
+func (v *NVar) parseContent(buf []byte) error {
+	// Try parsing as NVAR storage if it begins with NVAR signature
+	r := bytes.NewReader(buf)
+	var signature uint32
+	if err := binary.Read(r, binary.LittleEndian, &signature); err != nil {
+		return err
+	}
+	if signature != NVarEntrySignature {
+		return fmt.Errorf("NVAR Signature not found")
+	}
+	ns, err := NewNVarStore(buf)
+	if err != nil {
+		return fmt.Errorf("error parsing NVAR store in var %v: %v", v.Name, err)
+	}
+	v.NVarStore = ns
+	return nil
+}
+
+// newNVar parses a sequence of bytes and returns a NVar
+// object, if a valid one is passed, returns nil if buf is clear, or an error.
+func newNVar(buf []byte, offset uint64, s *NVarStore) (*NVar, error) {
+	v := NVar{Type: FullNVarEntry, Offset: offset}
+	// read the header and check for existing NVAR
+	if err := v.parseHeader(buf); err != nil {
+		// No NVAR check if remaining space is erased
+		if IsErased(buf, Attributes.ErasePolarity) {
```

Comment:
Can you move the `IsErased` check to the very beginning of this function?

Reply:
Right, in my mind it was more efficient to follow the list of headers instead of checking the full buffer for FFs, but you are right `IsErased` will stop really early if `buf` is not empty... so will do